File: leetcodePractice/copyBooks.py
def copyBooks(pages, k):
	if pages is None and len(pages) == 0:
		return 0
	
	if k == 0:
		return -1

	start, end = 0, sum(pages)

	while start + 1 < end:
		mid = (start + end) // 2
		if getNumberOfCopiers(pages, mid) <= k:
			end = mid
		else:
			start = mid

	if getNumberOfCopiers(pages, start) <= k:
		return start
	
	return end

def getNumberOfCopiers(pages, limit):
	copiers = 0
	lastCopied = limit

	for page in pages:
		if page > limit:
			return sum(pages)
		if lastCopied + page > limit:
			copiers += 1
		    lastCopied = 0

		lastCopied += page
	
	return copiers

File: leetcodePractice/7-15-2020/copyBooks.java
/**
 * 抄书问题
 */
public int copyBooks(int[] pages, int k) {
    if (pages == null || pages.length == 0) {
        return 0;
    }
    if (k == 0) {
        return -1;
    }

    int start = 0, end = Integer.MAX_VALUE;

    while (start + 1 < end) {
        int mid = start + (end - start) / 2;
        if (getNumberOfCopiers(pages, mid) <= k) {
            end = midl
        } else {
            start = mid;
        }
    }

    if (getNumberOfCopiers(pages, start) <= k) {
        return start;
    }
    return end;
}

private int getNumberOfCopiers(int[] pages, int limit) {
    int copiers = 0;
    int lastCopied = limit;

    for (int page : pages) {
        if (page > limit) {
            return Integer.MAX_VALUE;
        }
        if (lastCopied + page > limit) {
            copiers++;
            lastCopied = 0;
        }

        lastCopied += page;
    }
    return copiers;
}